src/main/java/build/buildfarm/server/OperationQueueService.java

```diff
@@ -63,6 +64,8 @@ public OperationQueueService(Instances instances) {
           responseObserver.onError(ex);
```

Comment:
I'm suddenly confused -- what can ever throw a StatusRuntimeException here? If it's the responseObserver itself, then calling its onError method will not work either.

In any case -- if we had an error here, why do we want to requeue the operation? Wouldn't it just be the exact same error the next time around?

This question applies in general to the whole premise of the requeing logic. If you want to do it as a retry layer inside Buildfarm, shouldn't you have a hard cap on them to not get into an infinite loop? And in general -- Bazel already has retries, so if you just return an INTERNAL_ERROR when something goes wrong on the server, Bazel will already retry with exponential decay (which is more likely to resolve the error) and likely hitting a different server and worker processes, which is also more likely to resolve the error. I'm wary of multiple nested retries, in general -- they often lead to way more executions than was expected.

Reply:
The SRE happens when the client has disconnected while waiting for a response. It only happens with CANCELLED status in this case, which is why I tested for all other results, and funnelled them to the responseObserver, because the caller's context (the instance) is not, strictly speaking, equipped or expecting to deal with grpc SREs - similar to doing one thing, it only expects to hear about success or failure to dispatch.

You're right about the requeue here though - It belongs back in the match implementation, which handles the long-lived operations that should redetermine whether a mismatched operation that fails to plaform or dispatch should be requeued.

Per the hard cap or fail fast mechanism consideration, the error recovery personality of bazel has obviously been fairly erratic, and the requeing logic functions more akin to error-correction for a worker-initiated requeue or a watcher requeue, or invariance for possibly long-period event sequences where a queue entry may no longer be valid (without having a worker to experimentally determine such). I would be concerned about setting a hard cap here where the requeues/op may become routine in, for instance, diverse operation/worker platform environments.

Reply:
Is there anything now that prevents reproducibly failing operations (from a future bug, for example) to be requeued forever? Would such an operation be dropped at any point?

Reply:
Nothing intrinsic to the requeuing behavior, no. The operation will be dropped systemically via expiration on the underlying cas delegate if there is sufficient inflow, and we're also obviously bound by the lifetime of the server itself.